File: similar/certguard-tls-rotation/environment/app/internal/report/report.go
// Package report renders certguard's artifacts: nginx rotation snippets, systemd
// override drop-ins, the canonical JSON trust report and the Markdown audit note.
package report

import (
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"certguard/internal/canon"
	"certguard/internal/decide"
)

const systemdOverride = "[Service]\nExecStart=\nExecStart=/bin/false\nNoNewPrivileges=yes\nProtectSystem=strict\n"

func fphex(fingerprint string) string {
	return strings.TrimPrefix(fingerprint, "sha256:")
}

func writeFile(outDir, rel, contents string) error {
	full := filepath.Join(outDir, filepath.FromSlash(rel))
	if err := os.MkdirAll(filepath.Dir(full), 0o755); err != nil {
		return err
	}
	return os.WriteFile(full, []byte(contents), 0o644)
}

func nullable(p *string) any {
	if p == nil {
		return nil
	}
	return *p
}

func chainArray(chain []string) []any {
	out := make([]any, 0, len(chain))
	for _, fp := range chain {
		out = append(out, fp)
	}
	return out
}

// WriteArtifacts writes every artifact for the evaluated services under outDir.
func WriteArtifacts(outDir, evaluatedAt string, results []decide.ServiceResult) error {
	services := make([]any, 0, len(results))
	rotations := make([]any, 0)
	blocked := make([]any, 0)

	for _, r := range results {
		services = append(services, map[string]any{
			"unit":                    r.Unit,
			"server_name":             r.ServerName,
			"enabled":                 r.Enabled,
			"bound_id":                r.BoundID,
			"decision":                r.Decision,
			"reason":                  r.Reason,
			"anchored":                r.Anchored,
			"bound_fingerprint":       nullable(r.BoundFingerprint),
			"bound_not_after":         nullable(r.BoundNotAfter),
			"chain":                   chainArray(r.Chain),
			"replacement_fingerprint": nullable(r.ReplacementFingerprint),
		})

		switch r.Decision {
		case "rotate":
			hex := fphex(*r.ReplacementFingerprint)
			snippet := "# certguard: rotated " + r.Unit + " (" + r.Reason + ")\n" +
				"ssl_certificate     /etc/certguard/certs/" + hex + ".pem;\n" +
				"ssl_certificate_key /etc/certguard/private/" + hex + ".key;\n"
			if err := writeFile(outDir, "nginx/"+r.Unit+".conf", snippet); err != nil {
				return err
			}
			rotations = append(rotations, map[string]any{"unit": r.Unit, "fingerprint": *r.ReplacementFingerprint})
		case "block":
			if err := writeFile(outDir, "systemd/"+r.Unit+"/override.conf", systemdOverride); err != nil {
				return err
			}
			blocked = append(blocked, r.Unit)
		}
	}

	report := map[string]any{
		"generated_by":   "certguard",
		"report_version": "1",
		"evaluated_at":   evaluatedAt,
		"services":       services,
		"rotations":      rotations,
		"blocked_units":  blocked,
	}
	if err := writeFile(outDir, "tls-trust-report.json", string(canon.WithNewline(report))); err != nil {
		return err
	}

	return writeFile(outDir, "tls-rotation-audit.md", renderMarkdown(evaluatedAt, results))
}

func renderMarkdown(evaluatedAt string, results []decide.ServiceResult) string {
	rotated, blocked, compliant := 0, 0, 0
	for _, r := range results {
		switch r.Decision {
		case "rotate":
			rotated++
		case "block":
			blocked++
		case "ok":
			compliant++
		}
	}
	lines := []string{
		"# TLS service rotation audit",
		"",
		"Evaluated at: " + evaluatedAt,
		"Services scanned: " + strconv.Itoa(len(results)),
		"Rotated: " + strconv.Itoa(rotated),
		"Blocked: " + strconv.Itoa(blocked),
		"Compliant: " + strconv.Itoa(compliant),
		"",
	}
	for _, r := range results {
		chainStr := "none"
		if len(r.Chain) > 0 {
			hexes := make([]string, 0, len(r.Chain))
			for _, fp := range r.Chain {
				hexes = append(hexes, fphex(fp))
			}
			chainStr = strings.Join(hexes, " > ")
		}
		var verdict string
		switch r.Decision {
		case "rotate":
			verdict = "ROTATED to " + fphex(*r.ReplacementFingerprint)
		case "block":
			verdict = "BLOCKED (" + r.Reason + ")"
		default:
			verdict = "COMPLIANT"
		}
		lines = append(lines,
			"## "+r.Unit,
			"",
			"- Server name: "+r.ServerName,
			"- Bound certificate: "+r.BoundID,
			"- Chain: "+chainStr,
			"- Decision: "+verdict,
			"",
		)
	}
	return strings.Join(lines, "\n")
}

File: similar/ffmpeg-service-hardening/environment/app/src/report.ts
// Artifact rendering.
//
// Writes the APT pin files, the systemd override drop-ins, the canonical JSON
// report and the Markdown audit note under the output directory.

import { mkdirSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { canonicalJson, compareCodePoints } from './canonical.js';
import { compareDeb } from './debver.js';
import type { ServiceDecision } from './policy.js';

export interface Pin {
  package: string;
  version: string;
}

function writeFile(outDir: string, relative: string, contents: string): void {
  const full = join(outDir, relative);
  mkdirSync(dirname(full), { recursive: true });
  writeFileSync(full, contents);
}

export function pinsFor(decisions: ServiceDecision[]): Pin[] {
  const targets = new Map<string, string>();
  for (const decision of decisions) {
    if (decision.decision === 'pin' && decision.package !== null && decision.pin_version !== null) {
      const current = targets.get(decision.package);
      if (current === undefined || compareDeb(decision.pin_version, current) > 0) {
        targets.set(decision.package, decision.pin_version);
      }
    }
  }
  return [...targets.keys()]
    .sort(compareCodePoints)
    .map((name) => ({ package: name, version: targets.get(name) as string }));
}

function renderMarkdown(decisions: ServiceDecision[]): string {
  const pinned = decisions.filter((d) => d.decision === 'pin').length;
  const blocked = decisions.filter((d) => d.decision === 'block').length;
  const compliant = decisions.filter((d) => d.decision === 'ok').length;
  const lines = [
    '# FFmpeg transcode hardening audit',
    '',
    `Services scanned: ${decisions.length}`,
    `Pinned: ${pinned}`,
    `Blocked: ${blocked}`,
    `Compliant: ${compliant}`,
    '',
  ];
  for (const decision of decisions) {
    const pkg = decision.package !== null ? decision.package : 'untracked';
    const version = decision.installed_version !== null ? decision.installed_version : 'unknown';
    let verdict: string;
    if (decision.decision === 'pin') verdict = `PINNED to ${decision.pin_version}`;
    else if (decision.decision === 'block') verdict = `BLOCKED (${decision.reason})`;
    else verdict = 'COMPLIANT';
    const advisories = decision.advisories.length > 0 ? decision.advisories.join(', ') : 'none';
    lines.push(
      `## ${decision.unit}`,
      '',
      `- Executable: ${decision.exec_path}`,
      `- Package: ${pkg}`,
      `- Installed version: ${version}`,
      `- Decision: ${verdict}`,
      `- Advisories: ${advisories}`,
      '',
    );
  }
  return lines.join('\n');
}

export function writeArtifacts(outDir: string, decisions: ServiceDecision[]): void {
  const ordered = [...decisions].sort((a, b) => compareCodePoints(a.unit, b.unit));
  const pins = pinsFor(ordered);
  const blocked = ordered.filter((d) => d.decision === 'block').map((d) => d.unit).sort(compareCodePoints);

  const report = {
    generated_by: 'ffguard',
    report_version: '1',
    services: ordered,
    pins,
    blocked_units: blocked,
  };
  writeFile(outDir, 'hardening-report.json', canonicalJson(report));

  // APT pin drop-ins.
  for (const pin of pins) {
    writeFile(
      outDir,
      join('apt', 'preferences', `${pin.package}.pref`),
      `Package: ${pin.package}\nPin: version ${pin.version}\nPin-Priority: 1001\n`,
    );
  }

  // systemd override drop-ins for blocked units.
  for (const decision of ordered) {
    if (decision.decision === 'block') {
      writeFile(
        outDir,
        join('systemd', decision.unit, 'override'),
        '[Service]\nExecStart=\nExecStart=/bin/false\nNoNewPrivileges=yes\nProtectSystem=strict\n',
      );
    }
  }

  writeFile(outDir, 'ffmpeg-hardening-audit.md', renderMarkdown(ordered));
}
